runtime/vm/resolvefield.cpp

```diff
@@ -578,6 +583,118 @@ addHiddenInstanceField(J9JavaVM *vm, const char *className, const char *fieldNam
+ *
+ * @param vm[in] pointer to the J9JavaVM
+ * @param reducedCpuUtil normalized cpu utilization of the hot field for the method being compiled
+ * @param clazz pointer to the class where a hot field should be added
+ * @param fieldOffset value of the field offset that should be added as a hot field for the given class
+ * @param reducedFrequency normalized block frequency of the hot field for the method being compiled
+ */
+void
+reportHotField(J9JavaVM *javaVM, int32_t reducedCpuUtil, J9Class* clazz, U_8 fieldOffset,  U_32 reducedFrequency)
+{
+	/* exit if pools are null */
+	if (NULL == clazz->classLoader->hotFieldPool || NULL == javaVM->hotFieldClassInfoPool || NULL == javaVM->hotFieldClassInfoPoolMutex) {
+		goto exit;
+	}
+
+	/* check if hotFieldsInfo exists already */
+	if(NULL == clazz->hotFieldsInfo) { 
+		omrthread_monitor_enter(javaVM->hotFieldClassInfoPoolMutex);
+		if(NULL == clazz->hotFieldsInfo) {
+        	/* initialize new hotFieldsInfo pool element as it does not exist already */
+			clazz->hotFieldsInfo = (J9ClassHotFieldsInfo *)pool_newElement(javaVM->hotFieldClassInfoPool);
+			if(NULL == clazz->hotFieldsInfo) {
+				omrthread_monitor_exit(javaVM->hotFieldClassInfoPoolMutex);
+				goto exit; 
+			}
+        	clazz->hotFieldsInfo->hotFieldListLength = 0;
+			clazz->hotFieldsInfo->consecutiveHotFieldSelections = 0;
+			clazz->hotFieldsInfo->hotFieldOffset1 = U_8_MAX;
+			clazz->hotFieldsInfo->hotFieldOffset2 = U_8_MAX;
+         
+        	/* create and initialize the new hot field */
+			omrthread_monitor_enter(clazz->classLoader->hotFieldPoolMutex);	
+			J9HotField* newHotField = (J9HotField *)pool_newElement(clazz->classLoader->hotFieldPool);	
+			if(NULL == newHotField) {
+				omrthread_monitor_exit(clazz->classLoader->hotFieldPoolMutex);
+				omrthread_monitor_exit(javaVM->hotFieldClassInfoPoolMutex);
+				goto exit; 
+			} 
+			newHotField->next = NULL;
+			newHotField->hotFieldOffset = fieldOffset;
+			newHotField->hotness = (reducedFrequency * reducedCpuUtil);
+			newHotField->cpuUtil = reducedCpuUtil;
+			omrthread_monitor_exit(clazz->classLoader->hotFieldPoolMutex);
+         
+        	/* update hotFieldsInfo for the class now that a new hot field has been added */
+			clazz->hotFieldsInfo->hotFieldListHead = newHotField;
+			clazz->hotFieldsInfo->isClassHotFieldListDirty = true;
+			clazz->hotFieldsInfo->hotFieldListLength++;
+			clazz->hotFieldsInfo->classLoader = clazz->classLoader;
+			omrthread_monitor_exit(javaVM->hotFieldClassInfoPoolMutex);
+			goto exit;		
+		} else {
+			omrthread_monitor_exit(javaVM->hotFieldClassInfoPoolMutex);
+		}
+	} 	
+
+	/* if the hot field exists, update the hot field with the newly reported hot field information */
+	J9HotField* previous;
+	J9HotField* current;
+	current = clazz->hotFieldsInfo->hotFieldListHead;
+	do {
+		if (current->hotFieldOffset == fieldOffset) {
+			updateHotField(current, clazz, reducedCpuUtil, reducedFrequency);
```

Comment:
This is what I meant about the duplication, `updateHotField` is essentially `addNewHotField` without the allocation of a hotfield. If in the future you need to update the metrics collected, it will have to be updated in two places.

Another approach would be to simply find the hotfield in this loop, once it is found then `break`. If it is not found then `current` is `NULL` which is fine. In line 659 check add a NULL check for `current` along with the existing check (youll need to acquire the monitor before hand). Then in line 660 where you have `addNewHotField`,  simply allocate a new hotfield. After the `if` block update the hotfield, the release the monitor. All of this can be done in a single function.

Reply:
Ok I will do something just like this 👍